### PR TITLE
[Bellesa] add new extractor

### PR DESCRIPTION
**Pull request by bellesafan**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] [Searched](https://github.com/ytdl-org/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Read [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site)
- [x] Read [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) and adjusted the code to meet them
- [x] Covered the code with tests (note that PRs without tests will be REJECTED)
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [x] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

Explanation of your *pull request* in arbitrary form goes here. Please make sure the description explains the purpose and effect of your *pull request* and is worded well enough to be understood. Provide as much context and examples as possible.

### CODE REVIEW

**Comment by dirkf** [comment]
As this doesn't have a linked site support request issue, please complete the site support request template: you can paste it here.

**Comment by bellesafan** [comment]
## Checklist



- [x] I'm reporting a new site support request
- [x] I've verified that I'm running youtube-dl version **2021.12.17**
- [x] I've checked that all provided URLs are alive and playable in a browser
- [x] I've checked that none of provided URLs violate any copyrights
- [x] I've searched the bugtracker for similar site support requests including closed ones


## Example URLs



- Single video: https://www.bellesa.co/videos/6698/quaranteened-pent-up


## Description



WRITE DESCRIPTION HERE

**Comment by bellesafan** [comment]
The website features nsfw content btw. Dunno if that's important to note.

**Comment by dirkf** [comment]
So I gathered. Thanks for your work, anyway.

Please
* add `# coding: utf-8` as the first line
* use `'single quoted strings'` rather than `"double"`
* explain or comment what the block `if "VideoCard" not in webpage:` is for.

Then I'll review in more detail.

**Comment by bellesafan** [comment]
I applied your suggestions. If there's anything else that needs adjustments, just let me know.

File: youtube_dl/extractor/bellesa.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+
+        webpage = self._download_webpage(
+            'https://www.bellesa.co/videos/%s/' % video_id, video_id)
+
+        # videos on this page are embedded into a container called VideoCard - if there is
+        # nothing on the page referencing a VideoCard we cannot extract the information and
+        # thus need to raise an error
+        if 'VideoCard' not in webpage:
```
Comment (dirkf) [inline youtube_dl/extractor/bellesa.py:40]: I suggest trying the extraction and then using this test (or, if the target is an element with attribute `id="VideoCard"`, `get_element_by_id('VideoCard', webpage)`) to generate a more specific diagnostic if no video is found.

File: youtube_dl/extractor/bellesa.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+            title = self._html_search_regex(
+                r'<title[^>]*>(?P<title>.+?)\s+\|\s+Bellesa',
+                webpage, 'title', default=None,
+                group='title', fatal=False)
+
+            raise ExtractorError('%s said: %s' % (self.IE_NAME, clean_html(title)), expected=True)
```
Comment (dirkf) [inline youtube_dl/extractor/bellesa.py:46]: Prefer:
```suggestion
            raise ExtractorError('[%s] %s: %s' % (self.IE_NAME, video_id, clean_html(title)), expected=True)
```
or define a new method
```py
    @classmethod
    def _raise_ExtractorError(cls, video_id, msg, expected=True):
        raise ExtractorError('[%s] %s: %s' % (cls.IE_NAME, video_id, msg or 'unspecified error', expected=expected)
```
and call it here and elsewhere.

File: youtube_dl/extractor/bellesa.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+        initial_data_raw = self._search_regex(r'window\.__INITIAL_DATA__\s+=\s+(.+?);</script>', webpage, 'initial_data')
```
Comment (dirkf) [inline youtube_dl/extractor/bellesa.py:48]: Whitespace may appear elsewhere, or not, and newlines may be in the `__INITIAL_DATA__`:
```suggestion
        initial_data_raw = self._search_regex(r'(?s)window\s*\.\s*__INITIAL_DATA__\s*=\s*(\{.+?\})\s*;\s*</script>', webpage, 'initial_data')
```

File: youtube_dl/extractor/bellesa.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+        try:
+            initial_data = json.loads(initial_data_raw)
+        except json.JSONDecodeError:
+            raise ExtractorError('%s said: cannot decode initial data', self.IE_NAME, expected=True)
```
Comment (dirkf) [inline youtube_dl/extractor/bellesa.py:53]: Use `_parse_json()` method.

File: youtube_dl/extractor/bellesa.py [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+
+        video = try_get(initial_data, lambda x: x['video'])
+        if not video:
+            raise ExtractorError('%s said: initial data malformed' % self.IE_NAME, expected=True)
+
+        resolutions = try_get(video, lambda x: x['resolutions'])
+        source = try_get(video, lambda x: x['source'])
```
Comment (dirkf) [inline youtube_dl/extractor/bellesa.py:60]: Simplify error check (after the 1st line we know `video` is a dict):
```suggestion
        video = try_get(initial_data, lambda x: x['video'], dict) or {}

        resolutions = video.get('resolutions')
        source = video.get('source')
```

File: youtube_dl/extractor/bellesa.py [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+
+        if not resolutions or not source:
+            raise ExtractorError('%s said: cannot extract playlist information from meta data' % self.IE_NAME, expected=True)
```
Comment (dirkf) [inline youtube_dl/extractor/bellesa.py:63]: As above.

File: youtube_dl/extractor/bellesa.py [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+
+        m3u8_url = 'https://s.bellesa.co/hls/v/%s/,%s,.mp4.urlset/master.m3u8' % (source, resolutions)
+        formats = self._extract_m3u8_formats(
+            m3u8_url, video_id, 'mp4',
+            entry_protocol='m3u8_native', m3u8_id='hls',
+            fatal=False)
+
+        self._sort_formats(formats)
```
Comment (dirkf) [inline youtube_dl/extractor/bellesa.py:71]: Insert diagnostic check before this, if desired:
```py
        if not formats:
            # if specific error, raise it
            ...
```

File: youtube_dl/extractor/bellesa.py [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+
+        # get from video meta data first
+        title = video.get('title')
+        if title:
+            title = title.strip()
+        else:
+            # fallback to og:title, which needs some treatment
+            title = self._og_search_title(webpage)
```
Comment (dirkf) [inline youtube_dl/extractor/bellesa.py:79]: In the error handling, you also used the `<title>` element. Now, it seems to be a defect that we don't have a generic title getter. As an author of `extractor/generic.py` wrote:
>        # it's tempting to parse this further, but you would
>        # have to take into account all the variations like
>        #   Video Title - Site Name
>        #   Site Name | Video Title
>        #   Video Title - Tagline | Site Name
>        # and so on and so forth; it's just not practical

I take this as a challenge, but not one that I've yet solved. Generally if `og:title` is available, it's a good choice if the title isn't in metadata.

File: youtube_dl/extractor/bellesa.py [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+            m3u8_url, video_id, 'mp4',
+            entry_protocol='m3u8_native', m3u8_id='hls',
+            fatal=False)
+
+        self._sort_formats(formats)
+
+        # get from video meta data first
+        title = video.get('title')
+        if title:
+            title = title.strip()
```
Comment (dirkf) [inline youtube_dl/extractor/bellesa.py:76]: Condition the value:
```suggestion
        title = strip_or_none(video.get('title'))
        if title:
```

File: youtube_dl/extractor/bellesa.py [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+        else:
+            # fallback to og:title, which needs some treatment
+            title = self._og_search_title(webpage)
+            if title:
+                title = title.split('|')[0].strip()
+
+        tags = None
+        tag_string = video.get('tags')
+        if tag_string:
+            tags = [c for c in map(lambda s: s.strip(), tag_string.split(','))]
```
Comment (dirkf) [inline youtube_dl/extractor/bellesa.py:86]: Simplify:
```suggestion
        tags = list(filter(None, map(lambda s: s.strip(), (video.get('tags') or '').split(','))))
```

File: youtube_dl/extractor/bellesa.py [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+
+        categories = None
+        if 'categories' in video:
+            categories = [c['name'] for c in video.get('categories')]
```
Comment (dirkf) [inline youtube_dl/extractor/bellesa.py:90]: And again:
```suggestion
        categories = None
        if 'categories' in video:
            categories = [c['name'] for c in video.get('categories')]
        list(filter(None, map(lambda d: strip_or_none(d['name']), (video.get('categories') or []))))
```

File: youtube_dl/extractor/bellesa.py [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+            'id': video_id,
+            'title': title,
+            'thumbnail': try_get(video, lambda x: x['image']),
+            'description': description,
+            'creator': try_get(video, lambda x: x['content_provider'][0]['name']),
+            'timestamp': try_get(video, lambda x: x['posted_on']),
+            'duration': try_get(video, lambda x: x['duration']),
+            'view_count': try_get(video, lambda x: x['views']),
+            'tags': tags,
+            'categories': categories,
```
Comment (dirkf) [inline youtube_dl/extractor/bellesa.py:106]: We know that `video` is a dict, but values should be conditioned:
```suggestion
        return {
            'id': video_id,
            'title': title,
            'thumbnail': url_or_none(video.get('image')),
            'description': strip_or_none(video.get('description')) or None,
            'creator': try_get(video, lambda x: x['content_provider'][0]['name'].strip(), compat_str),
            'timestamp': int_or_none(video.get('posted_on')),
            'duration': int_or_none(video.get('duration')),
            'view_count': int_or_none(video.get('views')),
            'tags': tags or None,
            'categories': categories or None,
```

**Review by dirkf — CHANGES_REQUESTED** [review]
Thanks, looking good but I've made a few suggestions.

File: youtube_dl/extractor/bellesa.py [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+            entry_protocol='m3u8_native', m3u8_id='hls',
+            fatal=False)
+
+        self._sort_formats(formats)
+
+        # get from video meta data first
+        title = video.get('title')
+        if title:
+            title = title.strip()
+        else:
```
Comment (dirkf) [inline youtube_dl/extractor/bellesa.py:77]: Condition the value:
```suggestion
        title = strip_or_none(video.get('title'))
        if not title:
```

File: youtube_dl/extractor/bellesa.py [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+    _VALID_URL = r'''https?://www\.bellesa\.co/videos/(?P<id>[0-9]+)/'''
+    _TESTS = [{
+        'url': 'https://www.bellesa.co/videos/2189/my-first-time-kissing-women',
+        'md5': '543b002f4d1e5b4a0dde243f4a43fd28',
+        'info_dict': {
+            'id': '2189',
+            'ext': 'mp4',
+            'title': 'My First Time Kissing Women',
+            'thumbnail': 'https://c.bellesa.co/dkvdbifey/image/upload/v1599024046/video_upload/2189cover.jpg',
+            'description': 'Jenna opens up about her troubles navigating a long-distance relationship. Her girlfriends ask if she’d ever be open to cheating on him but she says she would never, they’ve been together since college. Lena and Carter sit on either side of her and remind her it isn’t cheating if it’s with a girl…they start kissing and rubbing her chest, going down on her together. This girl on girl scene is so sensual, and the nerves on Jenna are real — exploring your sexuality takes courage, but there’s a lot of good vibes in this threesome.',
```
Comment (dirkf) [inline youtube_dl/extractor/bellesa.py:20]: With a long description, use MD5:
```suggestion
            'description': 'md5:69eea8a4ee31d42d6fd6302ad9e09ab2',
```

File: youtube_dl/extractor/bellesa.py (hunk 8 above)
Comment (bellesafan) [reply youtube_dl/extractor/bellesa.py:83]: I must admit I do understand whether or not I should change anything here

**Pull request closed without merging**